doc/neps/nep-0035-array-creation-dispatch-with-array-function.rst

```diff
@@ -8,16 +8,176 @@ NEP 35 — Array Creation Dispatching With __array_function__
+ultimately lead to an additional conversion into a downstream array type.
+
+Support for Python 2.7 has been dropped since NumPy 1.17, therefore we make use
+of the keyword-only argument standard described in PEP-3102 [2]_ to implement
+``like=``, thus preventing it from being passed by position.
+
+.. _neps.like-kwarg.usage-and-impact:
+
+Usage and Impact
+----------------
+
+To understand the intended use for ``like=``, and before we move to more complex
+cases, consider the following illustrative example consisting only of NumPy and
+CuPy arrays:
+
+.. code:: python
+
+    import numpy as np
+    import cupy
+
+    def my_pad(arr, padding):
+        padding = np.array(padding, like=arr)
+        return np.concatenate((padding, arr, padding))
+
+    my_pad(np.arange(5), [-1, -1])    # Returns np.ndarray
+    my_pad(cupy.arange(5), [-1, -1])  # Returns cupy.core.core.ndarray
+
+Note in the ``my_pad`` function above how ``arr`` is used as a reference to
+dictate what array type padding should have, before concatenating the arrays to
+produce the result. On the other hand, if ``like=`` wasn't used, the NumPy case
+case would still work, but CuPy wouldn't allow this kind of automatic
+conversion, ultimately raising a
+``TypeError: Only cupy arrays can be concatenated`` exception.
+
+Now we should look at how a library like Dask could benefit from ``like=``.
+Before we understand that, it's important to understand a bit about Dask basics
+and ensures correctness with ``__array_function__``. Note that Dask can compute
+different sorts of objects, like dataframes, bags and arrays, here we will focus
+strictly on arrays, which are the objects we can use ``__array_function__``
+with.
+
+Dask uses a graph computing model, meaning it breaks down a large problem in
+many smaller problems and merge their results to reach the final result. To
+break the problem down into smaller ones, Dask also breaks arrays into smaller
+arrays, that it calls "chunks". A Dask array can thus consist of one or more
+chunks and they may be of different types. However, in the context of
+``__array_function__``, Dask only allows chunks of the same type, for example,
+a Dask array can be formed of several NumPy arrays or several CuPy arrays, but
+not a mix of both.
+
+To avoid mismatched types during compute, Dask keeps an attribute ``_meta`` as
+part of its array throughout computation, this attribute is used to both predict
+the output type at graph creation time and to create any intermediary arrays
+that are necessary within some function's computation. Going back to our
+previous example, we can use ``_meta`` information to identify what kind of
+array we would use for padding, as seen below:
+
+.. code:: python
+
+    import numpy as np
+    import cupy
+    import dask.array as da
+    from dask.array.utils import meta_from_array
+
+    def my_pad(arr, padding):
+        padding = np.array(padding, like=meta_from_array(arr))
```

Comment:
This seems clunky... As a library writer, I want to *ignore* all cupy and dask code, but just have my library work seamlessly with them.

I guess I have to see how this is implemented but I would *assume* that `like=` passes the responsibility of array creation to downstream libraries. If so, couldn't we use `like=arr`, and leave it up to dask to give us an array primitive with the correct `_meta`?

What am I missing?

Reply:
I think this example is trying to solve the same problem as brought up in https://github.com/numpy/numpy/pull/17093/files#r472995422.

I agree though, I'd expect `like=arr` to just work here.

Reply:
Sorry, this is meant to explain how you would write the code for Dask as a Dask maintainer, and how it's expected to be written in a possible Dask implementation of that function. I do understand how it may look strange for a non-Dask developer, but I thought it would be useful to write code that would look real and work in their native environments -- in that case, Dask. I could remove it if you want and only leave the CuPy example, but today Dask is one of the largest users (if not _the_ largest) of `__array_function__`, so I thought I would clarify how that addresses a more complex case too.

> I think this example is trying to solve the same problem as brought up in https://github.com/numpy/numpy/pull/17093/files#r472995422.
> 
> I agree though, I'd expect `like=arr` to just work here.

It's not the same problem, it's just that here we want a CuPy array, which is the chunk type, not a new Dask array backed by CuPy.

Reply:
If this is a dask specific example, then calling the function `my_dask_pad` would make that intent clear.

Reply:
> If this is a dask specific example, then calling the function my_dask_pad would make that intent clear.

Fair enough, I did that change in https://github.com/numpy/numpy/pull/17093/commits/2e30534564aaf49cb0a12baf6b721b7b31402647 .